raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -2381,25 +2396,51 @@ public void complete() {
         }
     }
 
-    private final class ListenerContext implements CloseListener<BatchReader<T>> {
+    private static enum RegistrationOps {
+        REGISTER, UNREGISTER
+    }
+
+    private final class Registration {
+        private final RegistrationOps ops;
+        private final Listener<T> listener;
+
+        private Registration(RegistrationOps ops, Listener<T> listener) {
+            this.ops = ops;
+            this.listener = listener;
+        }
+
+        private void update(Map<Listener<T>, KafkaListenerContext> contexts) {
+            if (ops == RegistrationOps.REGISTER) {
+                if (contexts.putIfAbsent(listener, new KafkaListenerContext(listener)) != null) {
+                    logger.error("Attempting to add a listener that already exists: {}", listenerName(listener));
+                }
+            } else {
+                if (contexts.remove(listener) == null) {
+                    logger.error("Attempting to remove a listener context that doesn't exists: {}", listenerName(listener));
+                }
+            }
+        }
+    }
+
+    private final class KafkaListenerContext implements CloseListener<BatchReader<T>> {
```

Comment:
nit: is the `Kafka` prefix adding much value here?

Reply:
Done.